presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -85,441 +94,708 @@
+                    format("Corrupted partition statistics (Table: %s Partition: [%s]): %s", table, partition, format(message, args)));
         }
-        if (type.equals(REAL)) {
-            return intBitsToFloat(((Long) value).intValue());
+    }
+
+    private static TableStatistics getTableStatistics(
+            Map<String, ColumnHandle> columns,
+            Map<String, Type> columnTypes,
+            List<HivePartition> partitions,
+            Map<String, PartitionStatistics> statistics)
+    {
+        if (statistics.isEmpty()) {
+            return TableStatistics.empty();
         }
-        if (type instanceof DecimalType) {
-            DecimalType decimalType = (DecimalType) type;
-            if (isShortDecimal(decimalType)) {
-                return parseDouble(Decimals.toString((Long) value, decimalType.getScale()));
+
+        checkArgument(!partitions.isEmpty(), "partitions is empty");
+
+        OptionalDouble optionalRowsPerPartition = calculateRowsPerPartition(statistics.values());
+        if (!optionalRowsPerPartition.isPresent()) {
+            return TableStatistics.empty();
+        }
+        double rowsPerPartition = optionalRowsPerPartition.getAsDouble();
+        verify(rowsPerPartition >= 0, "rowsPerPartition must be greater than or equal to zero");
+        int queriedPartitionsCount = partitions.size();
+        double rowCount = rowsPerPartition * queriedPartitionsCount;
+
+        TableStatistics.Builder result = TableStatistics.builder();
+        result.setRowCount(Estimate.of(rowCount));
+        for (Map.Entry<String, ColumnHandle> column : columns.entrySet()) {
+            String columnName = column.getKey();
+            HiveColumnHandle columnHandle = (HiveColumnHandle) column.getValue();
+            Type columnType = columnTypes.get(columnName);
+            ColumnStatistics columnStatistics;
+            if (columnHandle.isPartitionKey()) {
+                columnStatistics = createPartitionColumnStatistics(columnHandle, columnType, partitions, statistics, rowsPerPartition, rowCount);
             }
-            if (isLongDecimal(decimalType)) {
-                return parseDouble(Decimals.toString((Slice) value, decimalType.getScale()));
+            else {
+                columnStatistics = createDataColumnStatistics(columnName, columnType, rowCount, statistics.values());
             }
-            throw new IllegalArgumentException("Unexpected decimal type: " + decimalType);
-        }
-        if (type.equals(DATE)) {
-            return (Long) value;
+            result.setColumnStatistics(columnHandle, columnStatistics);
         }
-        throw new IllegalArgumentException("Unsupported type: " + type);
+        return result.build();
     }
 
-    private OptionalDouble calculateRowsPerPartition(Map<String, PartitionStatistics> statisticsSample)
+    @VisibleForTesting
+    static OptionalDouble calculateRowsPerPartition(Collection<PartitionStatistics> statistics)
```

Comment:
nit: perhaps, add `Avg` to the name to clarify that this method computes average number of rows

Reply:
renamed to `calculateAverageRowsPerPartition`. Also renamed all the `rowsPerPartition` variables to `averageRowsPerPartition`